backend/memory/backend.go

```diff
@@ -44,7 +44,9 @@ func (be *Backend) Updates() <-chan backend.Update {
 }
 
 func (be *Backend) PushUpdate(update backend.Update) {
+	wait := update.Done()
 	be.updates <- update
+	<-wait
```

Comment:
Updates need to be processed (sent to the client's response channels) before the command has finished.  Otherwise the untagged update responses could come after the OK response for that command.

It might be better to keep a per-connection update queue that only gets flushed for some commands (IDLE, NOOP, NOTIFY, etc...).

Reply:
Got it.

>It might be better to keep a per-connection update queue that only gets flushed for some commands (IDLE, NOOP, NOTIFY, etc...).

That's a more complicated issue and I would like to leave it for another time, this PR is already pretty big. 

Various issues related to buffered EXPUNGE updates when mailbox is used by multiple clients are described in https://tools.ietf.org/html/rfc2180#section-4 and also briefly mentioned in https://tools.ietf.org/html/rfc2683#section-3.4.6. Current go-imap behavior does not break RFC 3501, strictly speaking, but creates a lot of troubles for clients.